src/app/fdctl/configure/ethtool-gro.c

```diff
@@ -0,0 +1,184 @@
+#include "configure.h"
+
+#include <stdio.h>
+#include <ctype.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+#include <sys/stat.h>
+#include <linux/if.h>
+#include <linux/ethtool.h>
+#include <linux/sockios.h>
+
+#define NAME "ethtool-gro"
+
+static int
+enabled( config_t * const config ) {
+  /* if we're running in a network namespace, we configure ethtool on
+      the virtual device as part of netns setup, not here */
+  return !config->development.netns.enabled;
+}
+
+static void
+init_perm( fd_caps_ctx_t *  caps,
+           config_t * const config ) {
+  (void)config;
+  fd_caps_check_root( caps, NAME, "disable network device generic-receive-offload (gro) with `ethtool --set-offload generic-receive-offload off`" );
+}
+
+static int
+device_is_bonded( const char * device ) {
+  char path[ PATH_MAX ];
+  FD_TEST( fd_cstr_printf_check( path, PATH_MAX, NULL, "/sys/class/net/%s/bonding", device ) );
+  struct stat st;
+  int err = stat( path, &st );
+  if( FD_UNLIKELY( err && errno != ENOENT ) )
+    FD_LOG_ERR(( "error checking if device `%s` is bonded, stat(%s) failed (%i-%s)",
+                 device, path, errno, fd_io_strerror( errno ) ));
+  return !err;
+}
+
+static void
+device_read_slaves( const char * device,
+                    char         output[ 4096 ] ) {
+  char path[ PATH_MAX ];
+  FD_TEST( fd_cstr_printf_check( path, PATH_MAX, NULL, "/sys/class/net/%s/bonding/slaves", device ) );
+
+  FILE * fp = fopen( path, "r" );
+  if( FD_UNLIKELY( !fp ) )
+    FD_LOG_ERR(( "error configuring network device, fopen(%s) failed (%i-%s)", path, errno, fd_io_strerror( errno ) ));
+  if( FD_UNLIKELY( !fgets( output, 4096, fp ) ) )
+    FD_LOG_ERR(( "error configuring network device, fgets(%s) failed (%i-%s)", path, errno, fd_io_strerror( errno ) ));
+  if( FD_UNLIKELY( feof( fp ) ) ) FD_LOG_ERR(( "error configuring network device, fgets(%s) failed (EOF)", path ));
```

Comment:
Why do we check that we are _not_ EOF after reading the line?

Reply:
This was in code copied from the original ethtool.c, which has been renamed ethtool-channels.c
I guess EOF is only set when an attempt is made to read beyond the end of file, but fgets only reads up to an including the '\n'. Perhaps it's meant to ensure we did reach the end of the bonded devices?